### PR TITLE
[pytorch][mobile] remove backward functions from jit-op-registry for mobile build

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26702 [pytorch][perf] add mobile friendly at:parallel_for backend
* #26746 [pytorch] fix ParallelNative.h/cpp build
* **#26657 [pytorch][mobile] remove backward functions from jit-op-registry for mobile build**
* #26700 [pytorch] rename caffe2::mobile_threadpool to caffe2::mobile_pthreadpool

Summary:
Add codegen option to remove backward ops from jit-op-registry as they are not
likely to be used for inference only mobile build.

Measured ARM-v7 AAR build size change: 5,804,182 -> 5,331,219.

Test Plan:
- build and integrate with demo app;

Differential Revision: [D17533887](https://our.internmc.facebook.com/intern/diff/D17533887)

### CODE REVIEW

File: tools/jit/gen_jit_dispatch.py [hunk 1]
```diff
@@ -238,6 +238,10 @@ def is_out_variant(decl):
     return decl['name'].endswith('_out')
 
 
+def is_backward_op(decl):
+    return decl['name'].endswith('_backward') or decl['name'].endswith('_backward_out')
```
Comment (ljk53) [inline tools/jit/gen_jit_dispatch.py]: This looks a bit hacky but there is an example above searching for "_out" and another example searching for "backward": https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L574

File: tools/setup_helpers/generate_code.py [hunk 2]
```diff
@@ -48,7 +48,11 @@ def generate_code(ninja_global=None,
             'tools/autograd',
             disable_autograd=disable_autograd,
         )
-        gen_jit_dispatch(declarations_path or DECLARATIONS_PATH, jit_gen_dir, 'tools/jit/templates')
+        gen_jit_dispatch(
+            declarations_path or DECLARATIONS_PATH,
+            jit_gen_dir,
+            'tools/jit/templates',
+            disable_autograd=disable_autograd)
```
Comment (ljk53) [inline tools/setup_helpers/generate_code.py]: I think technically backward ops are not coupled with autograd? Maybe I should rename all "disable_autograd" to "inference_only"? Just don't want to create too many different concepts while the main purpose is just to shrink prebuilt mobile lib size...

File: tools/jit/gen_jit_dispatch.py [hunk 3]
```diff
@@ -322,6 +326,14 @@ def emit_decl_variant(decl):
                                              lvalues=lvalues)
         return constructor
 
+    def filter_decls(jit_decls, disable_autograd):
```
Comment (ljk53) [inline tools/jit/gen_jit_dispatch.py]: I'm going to add another option to take a while list of ops to create custom mobile build so making this a separate function.

**Comment by ljk53** [comment]
Add @zdevito and @suo to help review.
@gchanan made a good point during offline discussion: we don't bind these backward functions in python, probably should simply skip all of them by default?
https://github.com/pytorch/pytorch/blob/128a65e2e0e60339b513426bc80226899ff94b51/tools/autograd/gen_python_functions.py#L21

File: tools/jit/gen_jit_dispatch.py (hunk 1 above)
Comment (dzhulgakov) [reply tools/jit/gen_jit_dispatch.py]: can we use the same set of filters as in python bindings? (and add a comment about it)

File: tools/setup_helpers/generate_code.py (hunk 2 above)
Comment (dzhulgakov) [reply tools/setup_helpers/generate_code.py]: I think the same setting makes sense - and the backward ops ARE coupled with autograd. They are not exposed to python, so the only way to use them is through autograd

**Review by dzhulgakov — COMMENTED** [review]
> @gchanan made a good point during offline discussion: we don't bind these backward functions in python, probably should simply skip all of them by default?

I think they should be still bound to JIT because of symbolic gradients

File: tools/jit/gen_jit_dispatch.py (hunk 1 above)
Comment (ljk53) [reply tools/jit/gen_jit_dispatch.py]: > can we use the same set of filters as in python bindings? (and add a comment about it)

Do you think I should use the entire SKIP_PYTHON_BINDINGS / even directly references it?

It has a lot other stuff like '.*_forward', 'index', .. - I haven't verified but do you think these can all be safely deleted from JIT binding?

**Comment by ljk53** [comment]
Seems I hit some weird bug with ghstack here - #26700 used to be after this PR, I landed it first as it doesn't really depend on this PR and swapped their order. It somehow marked this PR as merged.
I can recreate this PR but hope it doesn't cause any other problem...

cc: @ezyang

**Comment by ezyang** [comment]
Yeah just make a new PR. Sorry about the trouble. If you can send me some rages I can look into it.

**Pull request merged**